Create ConcurrencyRoadmap.md (#34517)
swiftlang/swift: docs/ConcurrencyRoadmap.md

Here is the above code rewritten using our proposed new syntax:

```swift
internal func refreshPlayers() async {
  players = await gameSession.allPlayers().map(\.nickname)
}
```

Things to note about this example:

- `refreshPlayers` is now an `async` function.
- `allPlayers` is also an `async` function, and _returns_ its result instead of passing it into a completion handler.
- Because of this, we can use expression composition to call the `map` function directly on the returned value.
- The `await` keyword appears before the expression calling `allPlayers`, indicating that the `refreshPlayers` function can become _suspended_ at this point.
- `await` works similarly to `try`, in that it need only appear once at the start of an expression that can suspend, not directly in front of every call that can suspend within that expression.

> - The explicit `self.` has been eliminated from property accesses, because no escaping closures capturing `self` are required.

Reviewer: > because no escaping closures capturing `self` are required.

because in escaping closures capturing `self` is required.

> - The accesses to the properties `allPlayers` and `players` now cannot have data races.

Reviewer 2: `allPlayers` is not a property.

> To understand how the the last point is achieved we must step out a layer and look at how queues are used to protect state.

Reviewer 3: ```suggestion
To understand how the last point is achieved we must step out a layer and look at how queues are used to protect state.
```

The original code was a method on a class that used `refreshQueue` to protect its internal state:

```swift
class PlayerRefreshController {
  var players: [String] = []
  var gameSession: GameSession
  var refreshQueue = DispatchQueue(label: "PlayerRefresh")
    
  func refreshPlayers(completion: (() -> Void)? = nil) { 
    ... 
  }
}
```

This is a common pattern: a class with a private queue and some properties that should only be accessed on the queue. We replace this manual queue management with an actor class:

```swift

> actor class PlayerRefreshControler {

Reviewer 4: ```suggestion
actor class PlayerRefreshController {
```

  var players: [String] = []
  var gameSession: GameSession

  func refreshPlayers() async { ... }
}
```

Things to note about this example:

- Declaring a class to be an actor is similar to giving a class a private queue and synchronizing all access to its private state through that queue.

> - Because this synchronization is now understood by the compiler, you cannot _forget_ to use the queue to protect state: the compiler will ensure that you are running on the queue in the class's methods, and it will prevent you from accessing the state outside those methods.

Reviewer 5: ```suggestion
- Because this synchronization is now understood by the compiler, you _won’t forget_ to use the queue to protect state: the compiler will ensure that you are running on the queue in the class's methods, and it will prevent you from accessing the state outside those methods.
```

> - Because this synchronization is now understood by the compiler, you cannot _forget_ to use the queue to protect state: the compiler will ensure that you are running on the queue in the class's methods, and it will prevent you from accessing the state outside those methods.

Reviewer 5: ```suggestion
- Because this synchronization is now understood by the compiler, you _won’t forget_ to use the queue to protect state: the compiler will ensure that you are running on the queue in the class's methods, and it will prevent you from accessing the state outside those methods.
```

- Because the compiler is responsible for doing this, it can be smarter about optimizing away synchronization, like when a method starts by calling an async function on a different actor.

Having this static relationship between the actor and its functions and properties lets us enforce the isolation of data to an actor and define away data races. We statically know whether we’re in a context that can safely access an actor’s properties, and if not, the compiler handles switching into that context.

Above we've shown an actor class, where you’ve got a tightly-encapsulated set of properties and code. But the way we do UI programming today often spreads code across a large number of classes that you’re supposed to use from a single main thread. That main thread is still a kind of actor — it’s what we call a global actor.

> You can mark classes and functions as being tied to that actor with an attribute. The compiler will let you reference this class from anywhere, but to actually call this method, you need to be on the UI actor. So, if it was appropriate for all the actions of `PlayerRefreshControler` to be performed on the global UI actor, we would represent it like this:

Reviewer 4: ```suggestion
You can mark classes and functions as being tied to that actor with an attribute. The compiler will let you reference this class from anywhere, but to actually call this method, you need to be on the UI actor. So, if it was appropriate for all the actions of `PlayerRefreshController` to be performed on the global UI actor, we would represent it like this:
```

```swift
@UIActor

> class PlayerRefreshControler {

Reviewer 4: ```suggestion
class PlayerRefreshController {
```

  var players: [String] = []
  var gameSession: GameSession
    
  func refreshPlayers() async {  ...  }
}
```

## Proposals for the First Phase

We will be pitching the following proposals in the coming weeks in support of the first phase:

- **`async`/`await`** introduces a [coroutine-based](https://en.wikipedia.org/wiki/Coroutine) `async`/`await` model to Swift. Functions will opt into to being `async`, and can then `await` the results of other `async` functions, allowing asynchronous code to be expressed in a more natural "straight-line" form. Pitch [here](https://forums.swift.org/t/concurrency-asynchronous-functions/41619), proposal [here](https://github.com/DougGregor/swift-evolution/blob/async-await/proposals/nnnn-async-await.md).

> - **`Task` API and Structured Concurrency** introduces the concept of a task to the standard library. This will cover APIs to create detached tasks, task "nurseries" for dynamically creating child tasks,  and the mechanics for cancellation and prioritization of tasks. It also introduces scope-based mechanisms for awaiting values from multiple child tasks, based on the principles of [structured concurrency](https://en.wikipedia.org/wiki/Structured_concurrency). Pitch [here](https://forums.swift.org/t/concurrency-structured-concurrency/41622), proposal [here](https://github.com/DougGregor/swift-evolution/blob/structured-concurrency/proposals/nnnn-structured-concurrency.md).

Reviewer 5: ```suggestion
- **`Task` API and Structured Concurrency** introduces the concept of a task to the standard library. This will cover APIs to create detached tasks, task "nurseries" for dynamically creating child tasks, and the mechanics for cancellation and prioritization of tasks. It also introduces scope-based mechanisms for awaiting values from multiple child tasks, based on the principles of [structured concurrency](https://en.wikipedia.org/wiki/Structured_concurrency). Pitch [here](https://forums.swift.org/t/concurrency-structured-concurrency/41622), proposal [here](https://github.com/DougGregor/swift-evolution/blob/structured-concurrency/proposals/nnnn-structured-concurrency.md).
```

> - **`Task` API and Structured Concurrency** introduces the concept of a task to the standard library. This will cover APIs to create detached tasks, task "nurseries" for dynamically creating child tasks,  and the mechanics for cancellation and prioritization of tasks. It also introduces scope-based mechanisms for awaiting values from multiple child tasks, based on the principles of [structured concurrency](https://en.wikipedia.org/wiki/Structured_concurrency). Pitch [here](https://forums.swift.org/t/concurrency-structured-concurrency/41622), proposal [here](https://github.com/DougGregor/swift-evolution/blob/structured-concurrency/proposals/nnnn-structured-concurrency.md).

Reviewer 5: ```suggestion
- **`Task` API and Structured Concurrency** introduces the concept of a task to the standard library. This will cover APIs to create detached tasks, task "nurseries" for dynamically creating child tasks, and the mechanics for cancellation and prioritization of tasks. It also introduces scope-based mechanisms for awaiting values from multiple child tasks, based on the principles of [structured concurrency](https://en.wikipedia.org/wiki/Structured_concurrency). Pitch [here](https://forums.swift.org/t/concurrency-structured-concurrency/41622), proposal [here](https://github.com/DougGregor/swift-evolution/blob/structured-concurrency/proposals/nnnn-structured-concurrency.md).
```

We intend to introduce the features described in this roadmap in two phases: first by introducing the ability to create async functions and actors; and second, by enforcing full actor isolation.

The basic idea of actor isolation is similar to the idea of [exclusive access to memory](https://github.com/apple/swift-evolution/blob/main/proposals/0176-enforce-exclusive-access-to-memory.md), and builds upon it. Swift’s concurrency design aims to provide an easy-to-use and composable approach to safe concurrency by starting from the natural isolation of actors and then using ownership as a complementary tool.

The problem of actor isolation reduces to the problem of ensuring that all ordinary mutable memory is accessed only by a particular actor or task.  This in turn reduces to an analysis of how the memory is accessed and who can access it in the first place. We can categorize memory into a few groups:

* Properties of an actor will be protected by the actor.
* Immutable memory (such as a `let` constant), local memory (such as a local variable that’s never captured), and value component memory (such as a properties of a struct, or an enum case), are already protected from data races.

> * Unsafe memory (such as an arbitrary allocation referenced by an `UnsafeMutablePointer`) is associated with an unsafe abstraction.  It’s actively undesirable to try to force these abstractions to be used safely, because these abstractions are meant to be usable to bypass safe language rules when necessary. Instead, we have to trust the programmer to use these correctly.

Reviewer 5: ```suggestion
* Unsafe memory (such as an arbitrary allocation referenced by an `UnsafeMutablePointer`) is associated with an unsafe abstraction. It’s actively undesirable to try to force these abstractions to be used safely, because these abstractions are meant to be usable to bypass safe language rules when necessary. Instead, we have to trust the programmer to use these correctly.
```

> * Unsafe memory (such as an arbitrary allocation referenced by an `UnsafeMutablePointer`) is associated with an unsafe abstraction.  It’s actively undesirable to try to force these abstractions to be used safely, because these abstractions are meant to be usable to bypass safe language rules when necessary. Instead, we have to trust the programmer to use these correctly.

Reviewer 5: ```suggestion
* Unsafe memory (such as an arbitrary allocation referenced by an `UnsafeMutablePointer`) is associated with an unsafe abstraction. It’s actively undesirable to try to force these abstractions to be used safely, because these abstractions are meant to be usable to bypass safe language rules when necessary. Instead, we have to trust the programmer to use these correctly.
```

* Global memory (such as a global or static variable) can in principle be accessed by any code anywhere, so is subject to data races.
* Class component memory can also be accessed from any code that hold a reference to the class. This means that while the _reference_ to the class may be protected by an actor, passing that reference between actors exposes its properties to data races. Note that this also references to classes held within value types when these are passsed between actors.

The goal of **full actor isolation** is to ensure that these last two categorizations are protected by default.

### First Phase: Basic Actor Isolation

The first phase introduces safety benefits. Users will be able to protect  global variables with global actors, and to protect class members by converting them to actor classes. Frameworks that require access on a particular queue can define a global actor and default protocols to it.

A number of important cases of actor isolation will be enforced in this stage:

```swift
actor class MyActor {
  let immutable: String = "42"
  var mutableArray: [String] = []

  func synchronousFunction() {

> 	  mutableArray += ["syncFunction called"]

Reviewer 6: ```suggestion
    mutableArray += ["syncFunction called"]
```

> * Class component memory can also be accessed from any code that hold a reference to the class. This means that while the _reference_ to the class may be protected by an actor, passing that reference between actors exposes its properties to data races. Note that this also references to classes held within value types when these are passsed between actors.

Reviewer 7: Typo "passsed", and I think that last sentence should start with "This also includes references to classes held within value types"

Author: ```suggestion
* Class component memory can also be accessed from any code that hold a reference to the class. This means that while the _reference_ to the class may be protected by an actor, passing that reference between actors exposes its properties to data races. This also includes references to classes held within value types, when these are passed between actors.
```

> The first phase introduces safety benefits. Users will be able to protect  global variables with global actors, and to protect class members by converting them to actor classes. Frameworks that require access on a particular queue can define a global actor and default protocols to it.

Reviewer 5: ```suggestion
The first phase introduces safety benefits. Users will be able to protect global variables with global actors, and to protect class members by converting them to actor classes. Frameworks that require access on a particular queue can define a global actor and default protocols to it.
```

> The first phase introduces safety benefits. Users will be able to protect  global variables with global actors, and to protect class members by converting them to actor classes. Frameworks that require access on a particular queue can define a global actor and default protocols to it.

Reviewer 5: ```suggestion
The first phase introduces safety benefits. Users will be able to protect global variables with global actors, and to protect class members by converting them to actor classes. Frameworks that require access on a particular queue can define a global actor and default protocols to it.
```

  }
}

extension MyActor {

  func asyncFunction(otherActor: MyActor) async {
    // allowed: an actor can access its internal state, even in an extension
    self.mutableArray += ["asyncFunction called"]
  
    // allowed: immutable memory can be accessed from outside the actor
    print(other.immutable)

    // error: an actor cannot access another's mutable state
    otherActor.mutableArray += ["not allowed"]

    // error: either reading or writing
    print(other.mutableArray.first)
    
    // allowed: async functions can call async functions on other actors

>     other.asyncFunction(otherActor: self)

Reviewer 6: ```suggestion
    await other.asyncFunction(otherActor: self)
```

>     // error: only asynchronous functions can be called from outside the actor

Reviewer: > only asynchronous functions can be called **from outside** the actor

only asynchronous functions can be called **inside** the actor

Reviewer 8: No, the phrasing is correct here.

This is attempting to call a synchronous function on the `other` actor -- this we are outside of it.

>   func asyncFunction(otherActor: MyActor) async {
>     // allowed: an actor can access its internal state, even in an extension
>     self.mutableArray += ["asyncFunction called"]
>   
>     // allowed: immutable memory can be accessed from outside the actor
>     print(other.immutable)
> 
>     // error: an actor cannot access another's mutable state
>     otherActor.mutableArray += ["not allowed"]
> 
>     // error: either reading or writing
>     print(other.mutableArray.first)
>     
>     // allowed: async functions can call async functions on other actors
>     await other.asyncFunction(otherActor: self)
>     
>     // error: only asynchronous functions can be called from outside the actor
>     other.synchronousFunction()    
>   }

Reviewer 9: Fix variable name
```suggestion
  func asyncFunction(otherActor: MyActor) async {
    // allowed: an actor can access its internal state, even in an extension
    self.mutableArray += ["asyncFunction called"]
  
    // allowed: immutable memory can be accessed from outside the actor
    print(otherActor.immutable)

    // error: an actor cannot access another's mutable state
    otherActor.mutableArray += ["not allowed"]

    // error: either reading or writing
    print(otherActor.mutableArray.first)
    
    // allowed: async functions can call async functions on other actors
    await otherActor.asyncFunction(otherActor: self)
    
    // error: only asynchronous functions can be called from outside the actor
    otherActor.synchronousFunction()    
  }
```

Author: ```suggestion
  func asyncFunction(other: MyActor) async {
    // allowed: an actor can access its internal state, even in an extension
    self.mutableArray += ["asyncFunction called"]
  
    // allowed: immutable memory can be accessed from outside the actor
    print(other.immutable)

    // error: an actor cannot access another's mutable state
    otherActor.mutableArray += ["not allowed"]

    // error: either reading or writing
    print(other.mutableArray.first)
    
    // allowed: async functions can call async functions on other actors
    await other.asyncFunction(otherActor: self)
    
    // error: only asynchronous functions can be called from outside the actor
    other.synchronousFunction()    
  }
```

}
```

These enforcements are *non source breaking* because actors and async functions are a new feature.

### Second Phase: Full Actor Isolation

Even after the introduction of actors, there will still exist the possibility for race conditions, through global variables and values of reference type:

```swift
var racyGlobal: [String] = []

@MyGlobalActor

> Var safeGlobal: [String] = []

Reviewer 6: ```suggestion
var safeGlobal: [String] = []
```

>     otherActor.mutableArray += ["not allowed"]

Reviewer 10: should this be just `other`?
```suggestion
    other.mutableArray += ["not allowed"]
```

class PlainOldClass {
  var unprotectedState: String = []
}

> actor RacyActor {

Reviewer 6: ```suggestion
actor class RacyActor {
```

>     await other.asyncFunction(otherActor: self)

Reviewer 11: should be just `other` too?
```suggestion
    await other.asyncFunction(other: self)
```

  let immutableClassReference: PlainOldClass

  func racyFunction(other: RacyActor) async {
    // protected: global variable protected by a global actor
    safeGlobal += ["Safe access"]
  
    // unprotected: global variable not in an actor
    racyGlobal += ["Racy access"]
    
    // unprotected: racyProperty is immutable, but it is a reference type
    // so it allows access to unprotected shared mutable type
    other.takeClass(immutableClassReference)
  }
  

>   func takeClass(_ class: PlainOldClass) {

Author: ```suggestion
  func takeClass(_ plainClass: PlainOldClass) {
```

>     class.unprotectedState += ["Racy access"]  

Reviewer 6: ```suggestion
    `class`.unprotectedState += ["Racy access"]  
```

Author: ```suggestion
    plainClass.unprotectedState += ["Racy access"]  
```

>     // unprotected: racyProperty is immutable, but it is a reference type

Reviewer 3: Should this be "immutableClassReference"?
```suggestion
    // unprotected: immutableClassReference is immutable, but it is a reference type
```

  }
}
```

In the first phase, we intend to preserve Swift's *current* default behavior: global variables and class component memory are not protected from data races. "Actor unsafe" is therefore the default for this memory. Because this is the default of current Swift, enabling this first phase is *non-source breaking*.

In the second phase, the introduction of further features will provide a full set of tools for working with fully isolated actors. The most important of these is the ability to restrict a type to be "actor local". When a type is marked actor local, the compiler will prevent it from being passed between actors. Instead, the reference would need to be somehow cloned/unshared before being passed across a boundary.

This in turn will allow the defaults to be changed:

- Global variables will be required to be protected by a global actor, or marked "actor unsafe".

> - Classes (and types that contain class references) will change from being from being "actor unsafe" by default to being "actor local".

Reviewer 12: Fixed `from being from being`
```suggestion
- Classes (and types that contain class references) will change from being "actor unsafe" by default to being "actor local".
```

This change in default will necessitate a **source break**, and will need to be gated by a language mode. Code that touches a mutable global variable or shares a class reference across actor boundaries fundamentally cannot be shown to be safe from data races, and will need to change to ensure that it (and code written in the future) is safe from data races. It is hoped that this source break will not be onerous:

- it is expected that use of global variables should be sparing, and most global variables can be protected by a global actor;

> - as long as classes are not shared across actor boundaries, "actor local" annotation should  not affect code within an actor;

Reviewer 5: ```suggestion
- as long as classes are not shared across actor boundaries, "actor local" annotation should not affect code within an actor;
```

> - as long as classes are not shared across actor boundaries, "actor local" annotation should  not affect code within an actor;

Reviewer 5: ```suggestion
- as long as classes are not shared across actor boundaries, "actor local" annotation should not affect code within an actor;
```

- where references must be passed across boundaries, the language should make this obvious and the solutions simple;
- by further encouraging and simplifying the use of value types, the need to share classes across actor boundaries should be reduced;
- the time between the two phases will give users an opportunity to factor their code into actors and async functions, preparing it for full isolation.

Unlike the pitches for the first phase, which will be pitched and proposed, the language features needed for the second phase will initially be raised for _discussion_ on the evolution discussion section of the Swift forums. One of the main drivers for the two-phase approach is that of wanting to give Swift users time to become accustomed to async functions and actors before moving to a full isolation model. It is expected that the experience of porting large working code bases to actors and async functions will inform the functionality needed to enforce full actor isolation. This feedback should inform the discussion of features in the second phase.

> Features expected to be discussed as part of a second phase include:

Reviewer 5: ```suggestion
Features expected to be discussed as part of the second phase include:
```

> Features expected to be discussed as part of a second phase include:

Reviewer 5: ```suggestion
Features expected to be discussed as part of the second phase include:
```

- introduction of an `actorlocal` restriction on types;

> - compiler support for guaranteed correct "copy on write" types via a `mutableIfUnique` class type;

Reviewer 5: ```suggestion
- compiler support for guaranteed correct copy-on-write types via a `mutableIfUnique` class type;
```

> - compiler support for guaranteed correct "copy on write" types via a `mutableIfUnique` class type;

Reviewer 5: ```suggestion
- compiler support for guaranteed correct copy-on-write types via a `mutableIfUnique` class type;
```

- attributes to opt-out of actor isolation, for example where thread safety is being handled through some other means.

## Glossary of Concepts

These are the basic concepts that will be used throughout the design, briefly defined here.

* A **synchronous function** is the kind of function that Swift programmers are already used to: it runs to completion on a single thread, with no interleaving code except for any synchronous functions it calls.
* A **thread** refers to the underlying platform’s concept of threads. Platforms vary, but tend to share basic characteristics: true concurrency requires creating a platform thread, but creating and running platform threads is expensive. C function calls, and ordinary synchronous Swift functions, require the use of a platform thread.
* An **asynchronous function** is a new kind of function which does not have to run to completion without interruption. An interruption causes the function to be **suspended**. The point at which an async function may give up its thread is a **suspension point**.

> * A **task** is an operation that runs asynchronously. All asynchronous functions run as part of some task.  When an asynchronous function calls another asynchronous function, the call is still part of the the same task, even if the call has to change actors. A task is the analogue of a thread for asynchronous functions.

Reviewer 3: ```suggestion
* A **task** is an operation that runs asynchronously. All asynchronous functions run as part of some task.  When an asynchronous function calls another asynchronous function, the call is still part of the same task, even if the call has to change actors. A task is the analogue of a thread for asynchronous functions.
```

> * A **task** is an operation that runs asynchronously. All asynchronous functions run as part of some task.  When an asynchronous function calls another asynchronous function, the call is still part of the the same task, even if the call has to change actors. A task is the analogue of a thread for asynchronous functions.

Reviewer 5: ```suggestion
* A **task** is an operation that runs asynchronously. All asynchronous functions run as part of some task. When an asynchronous function calls another asynchronous function, the call is still part of the the same task, even if the call has to change actors. A task is the analogue of a thread for asynchronous functions.
```

> * A **task** is an operation that runs asynchronously. All asynchronous functions run as part of some task.  When an asynchronous function calls another asynchronous function, the call is still part of the the same task, even if the call has to change actors. A task is the analogue of a thread for asynchronous functions.

Reviewer 5: ```suggestion
* A **task** is an operation that runs asynchronously. All asynchronous functions run as part of some task. When an asynchronous function calls another asynchronous function, the call is still part of the the same task, even if the call has to change actors. A task is the analogue of a thread for asynchronous functions.
```

> * An async function can create a **child task**. Child tasks inherit some of the structure of their parent task, including its priority, but can run concurrently with it.  However, this concurrency is bounded: a function that creates a child task must wait for it to end before returning.

Reviewer 5: ```suggestion
* An async function can create a **child task**. Child tasks inherit some of the structure of their parent task, including its priority, but can run concurrently with it. However, this concurrency is bounded: a function that creates a child task must wait for it to end before returning.
```

> * An async function can create a **child task**. Child tasks inherit some of the structure of their parent task, including its priority, but can run concurrently with it.  However, this concurrency is bounded: a function that creates a child task must wait for it to end before returning.

Reviewer 5: ```suggestion
* An async function can create a **child task**. Child tasks inherit some of the structure of their parent task, including its priority, but can run concurrently with it. However, this concurrency is bounded: a function that creates a child task must wait for it to end before returning.
```

* A program wishes to initiate independent concurrent work that can outlast its spawning context uses a **detached task** rather than a bounded child task.
* A **partial task** is a unit of schedulable work. When the currently-executing function in a task is suspended, that is the end of the partial task, and a new partial task is created to continue the overall task’s work.
* An **executor** is a service which accepts the submission of partial tasks and arranges for some thread to run them. An async function that is currently running always knows the executor on which it's running. An executor is called **exclusive** if the partial tasks submitted to it will never be run concurrently.

> * An **actor** is an independent part of the program which can run code.  It can only run one piece of code at a time — that is, it acts as an exclusive executor — but the code it runs can execute concurrently with the code run by other actors.

Reviewer 5: ```suggestion
* An **actor** is an independent part of the program which can run code. It can only run one piece of code at a time — that is, it acts as an exclusive executor — but the code it runs can execute concurrently with the code run by other actors.
```

> * An **actor** is an independent part of the program which can run code.  It can only run one piece of code at a time — that is, it acts as an exclusive executor — but the code it runs can execute concurrently with the code run by other actors.

Reviewer 5: ```suggestion
* An **actor** is an independent part of the program which can run code. It can only run one piece of code at a time — that is, it acts as an exclusive executor — but the code it runs can execute concurrently with the code run by other actors.
```

> * An actor can have protected state which can only be accessed by that actor. The system through which this is achieved is called **actor isolation**. The long term goal is for Swift to guarantee actor isolation by default.

Reviewer 5: ```suggestion
* An actor can have protected state which can only be accessed by itself. The system through which this is achieved is called **actor isolation**. The long term goal is for Swift to guarantee actor isolation by default.
```

> * An actor can have protected state which can only be accessed by that actor. The system through which this is achieved is called **actor isolation**. The long term goal is for Swift to guarantee actor isolation by default.

Reviewer 5: ```suggestion
* An actor can have protected state which can only be accessed by itself. The system through which this is achieved is called **actor isolation**. The long term goal is for Swift to guarantee actor isolation by default.
```

* An **actor class** is a reference type, each instance of which is a separate actor. Its protected state is its instance properties, and its actor functions are its instance methods.

> * A **global actor** is a global object. Its protected state and actor functions may be spread across many different types.  They can be marked with an actor-specific attribute, which Swift can infer in many cases.

Reviewer 5: ```suggestion
* A **global actor** is a global object. Its protected state and actor functions may be spread across many different types. They can be marked with an actor-specific attribute, which Swift can infer in many cases.

```

> * A **global actor** is a global object. Its protected state and actor functions may be spread across many different types.  They can be marked with an actor-specific attribute, which Swift can infer in many cases.

Reviewer 5: ```suggestion
* A **global actor** is a global object. Its protected state and actor functions may be spread across many different types. They can be marked with an actor-specific attribute, which Swift can infer in many cases.

```
